### PR TITLE
ivy.el: Drop preselected candidate after input

### PR DESCRIPTION
**Pull request by raxod502**
Suggested fix for https://github.com/abo-abo/swiper/issues/1563.

### CODE REVIEW

File: ivy.el [hunk 1]
```diff
@@ -3066,6 +3066,20 @@ RE-STR is the regexp, CANDS are the current candidates."
...
+             (not (or
+                   (and (integerp (ivy-state-preselect ivy-last))
+                        (= ivy--index
+                           (ivy-state-preselect ivy-last)))
+                   (equal (ivy-state-current ivy-last)
+                          (ivy-state-preselect ivy-last))
+                   (and (stringp (ivy-state-preselect ivy-last))
+                        (string-match-p
+                         (ivy-state-preselect ivy-last)
+                         (ivy-state-current ivy-last)))))
```
Comment (basil-conto) [inline ivy.el]: Nitpick: might be nice to bind `ivy-state-preselect` and `ivy-state-current` to reduce visual clutter (and avoid duplicate calls).

Comment (basil-conto) [reply ivy.el]: (Seeing as there are other calls to these accessors in the function as well.)

File: ivy.el [hunk 2]
```diff
@@ -3066,6 +3066,18 @@ RE-STR is the regexp, CANDS are the current candidates."
              (not (and ivy--flx-featurep
                        (eq ivy--regex-function 'ivy--regex-fuzzy)
                        (< (length cands) 200)))
+             ;; If there was a preselected candidate, don't try to
+             ;; keep it selected even if the regexp still matches it.
+             ;; See issue #1563.  See also `ivy--preselect-index',
+             ;; which this logic roughly mirrors.
+             (not (let ((preselect (ivy-state-preselect ivy-last))
+                        (current (ivy-state-current ivy-last)))
```
Comment (basil-conto) [inline ivy.el]: Nitpick: again, there are other parts of `ivy--recompute-index` which could make use of these bindings.

**Comment by raxod502** [comment]
@basil-conto Fixed, thanks.

**Comment by raxod502** [comment]
@abo-abo Any comments?

**Comment by abo-abo** [comment]
Merged, thanks.

**Pull request closed without merging**